archivesunleashed/aut
# Change Id generation for graphs from using hashes for urls to using .zipWithUniqueIds()

**GitHub issue(s)**:

#243

# What does this Pull Request do?

This PR adds WriteGraph which generates GEXF and/or Graphml files.
There are also some additional utilities such as an id lookup.

This object generates proper unique ids. This differs from the current method in WriteGraphML and WriteGEXF that simply creates ids using an MD5 hash of the url.  This method is better because the latter method could produce incorrect graphs in the case where hashes collide (eg. with very large graphs). 

Timing tests with a medium-sized graph shows that it increases the processing time by 10-15%. However, it could reduce the size of network graph derivatives by an unknown margin (because numeric ids are smaller than hashes).

Example:
Old way produces:
```
<node id="405d19a958ba43d88e9edd7a77338aa3" label="laws.justice.gc.ca" />
<node id="69bde2cbb357119a50a950fca99a8341" label="english.uvic.ca" />
<node id="8687e4dc11548a5917504975feb7c649" label="oipc.bc.ca" />
<node id="30192bae759dafccc58bccc268c2b411" label="accuweather.com" />
```
New way:
```
<node id="0" label="laws.justice.gc.ca" />
<node id="38" label="english.uvic.ca" />
<node id="76" label="oipc.bc.ca" />
<node id="114" label="accuweather.com" />
```
# How should this be tested?

- Travis should pass.

```
def timed(f: => Unit) = {
  val start = System.currentTimeMillis()
  f
  val end = System.currentTimeMillis()
  println("Elapsed Time: " + (end - start))
}

timed {
import io.archivesunleashed._
import io.archivesunleashed.app._
import io.archivesunleashed.matchbox._

val links = RecordLoader.loadArchives("/Users/ryandeschamps/warcs/", sc)
  .keepValidPages()
  .map(r => (r.getCrawlDate, ExtractLinks(r.getUrl, r.getContentString)))
  .flatMap(r => r._2.map(f => (r._1, ExtractDomain(f._1).replaceAll("^\\s*www\\.", ""), ExtractDomain(f._2).replaceAll("^\\s*www\\.", ""))))
  .filter(r => r._2 != "" && r._3 != "")
  .countItems()
  .filter(r => r._2 > 5)

WriteGraph(links, "new-gephi3.gexf") }
```

Should produce a Gexf that opens in Gephi.

Change last line to ```WriteGraph.asGraphml(links, "new-gephi3.gexf")``` to get graphml.

I have not tested it in GraphPass yet, but there is no reason why it should not work as directed.

# Additional Notes:

- `WriteGraphml`would replace WriteGEXF & WriteGraphml which can be deprecated.
- `CommandLineApp` would also have to be changed before WriteGEXF etc. are removed.
- I think aut had a previous `WriteGraph` udf which was deprecated and removed.
- This applies only to the RDD graph functions.  DF functions are not changed. (I need to review the way DFs do unique ids)
- I added a node id lookup tool during the process of developing WriteGraph. It might be useful for the toolkit, but it can also be removed.  

# Interested parties

@lintool @ruebot @ianmilligan1 

Thanks in advance for your help with the Archives Unleashed Toolkit!
